### ISSUE TITLE
"type ... too big for the current architecture" error reported without span pointing to type

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust#25116]
e053dfad23515f7020171ae18013b230531a6042 added checks for huge types. These checks are done in middle/trans, but the location of the type is not reported.

For anonymous array types there's no way to extract the span, as the ast-ID is lost.

An example:

``` rust
fn main() {
    let _x = [0; 9223372036854775808];
}
```

the error message:

```
error: the type `[i32; 9223372036854775808]` is too big for the current architecture
```

**Comment by steveklabnik** [comment]
Triage: no change.

**Comment by czipperz** [comment]
Playground link: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=811f39850b75fc7334b05256223fdb97

**Comment by oli-obk** [comment]
The diagnostic is actually wrong. It's not about the type, it's about values of that type. You can use any type no matter its size for trait resolution and such, you just can't create values of it.

Additionally, we're still not emitting this error with spans. I think the only way to reasonably do this is to create a new well formedness bound that is *automatically* added to a function's where bounds for all types within that function that are used with values, but for which we can't prove well-formedness. This may be *very* expensive, since it spawns lots of new `where` bounds where there were none before. Not sure, but may be worth an experiment

**Comment by estebank** [comment]
Current output:

```
error: values of the type `[i32; 9223372036854775808]` are too big for the current architecture
```

**Comment by estebank** [comment]
There are a couple of places where we still don't have spans. One of them is when doing vtables, and that machinery doesn't track spans at all. We also don't use `spanned_layout_of` on as many places as we could.

**Comment by estebank** [comment]
Closed by accident. A PR description of "doesn't close #XXXX" got interpreted by GH as "close #XXXX" 😐
